File: jetbrains-plugin/src/main/kotlin/com/oneide/services/RuleService.kt
package com.oneide.services

import com.intellij.notification.NotificationAction
import com.intellij.notification.NotificationGroupManager
import com.intellij.notification.NotificationType
import com.intellij.openapi.Disposable
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.ApplicationNamesInfo
import com.intellij.openapi.application.WriteAction
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.openapi.util.Disposer
import com.oneide.models.Role
import com.oneide.utils.Debouncer
import com.oneide.utils.Logger
import java.io.File
import java.util.*

/**
 * Represents a rule file with its content and metadata.
 * @property path Relative path or filename.
 * @property content Content of the file.
 * @property lastModified Last modified timestamp.
 */
data class RuleFile(val path: String, val content: String, val lastModified: Long = 0)

data class RuleStat(val ai: String, val file: File, val mtime: Long)

/**
 * Interface for building rule files for a target AI tool.
 */
interface RuleBuilder {
    fun buildRules(sourceFiles: List<RuleFile>, sourceAi: String): List<RuleFile>
}

/**
 * Builds rules for tools that support folder-based rule structure.
 *
 * @property ruleRoot The root directory for the rules.
 */
class FolderRuleBuilder(private val ruleRoot: String, private val extension: String? = null) : RuleBuilder {
    override fun buildRules(sourceFiles: List<RuleFile>, sourceAi: String): List<RuleFile> {
        return sourceFiles.map { file ->
            var name = File(file.path).name

            if (extension != null) {
                val ext = File(name).extension
                if (ext.isNotEmpty() && !name.endsWith(extension)) {
                    name = name.substring(0, name.length - ext.length - 1) + extension
                } else if (ext.isEmpty()) {
                    name += extension
                }
            }
            // name = "${sourceAi.lowercase()}_$name" // Keep original filename

            val path = if (ruleRoot.isEmpty()) name else "$ruleRoot/$name"
            RuleFile(path, file.content, file.lastModified)
        }
    }
}

/**
 * Builds rules for tools that require a single file configuration.
 * Concatenates all source rules into one file.
 *
 * @property targetPath The path of the single target file.
 */
class SingleFileRuleBuilder(private val targetPath: String) : RuleBuilder {
    override fun buildRules(sourceFiles: List<RuleFile>, sourceAi: String): List<RuleFile> {
        // Iterate through each source file and extract its content
        val sb = StringBuilder()
        var maxTime = 0L
        for (file in sourceFiles) {
            sb.append(file.content).append("\n\n")
            if (file.lastModified > maxTime) maxTime = file.lastModified
        }
        // Return the single target file with the merged content
        return listOf(RuleFile(targetPath, sb.toString(), maxTime))
    }
}

/**
 * Represents the synchronization state of a rule set.
 *
 * @property source The name of the AI tool that is the source of the rules.
 * @property lastModified The timestamp when the rules were last modified.
 * @property isFromSynced Indicates if the current state is a result of a sync operation.
 * @property syncedTo List of AI tools that this rule set has been synced to.
 */
data class SyncState(
    var source: String? = null,
    var lastModified: Long = 0,
    var isFromSynced: Boolean = false,
    val syncedTo: MutableList<String> = mutableListOf()
)

/**
 * Service responsible for synchronizing AI rules between different tools.
 * It triggers a synchronization check only when this IDE instance becomes cluster Leader.
 *
 * @property project The IntelliJ project.
 */
class RuleService(
    private val project: Project,
    private val clusterService: ClusterService
) {
    private val debouncer = Debouncer(1000)
    private val currentAppName = ApplicationNamesInfo.getInstance().fullProductName
    private val logger = Logger.withProject(project)
    private var unsubscribeRoleChange: (() -> Unit)? = null

    /**
     * Starts the RuleService.
     * Triggers synchronization only when this instance becomes the cluster Leader.
     */
    fun start() {
        val configService = ConfigService.getInstance(project)
        if (!configService.getConfig().syncRules) {
            logger.info("AI Rule Sync is disabled.")
            return
        }

        logger.info("Starting RuleService for project: ${project.name}. App: $currentAppName")
        unsubscribeRoleChange = clusterService.addRoleChangeListener { role ->
            if (role != Role.LEADER) return@addRoleChangeListener
            debouncer.debounce { checkAndSync() }
        }
        Disposer.register(project, Disposable {
            unsubscribeRoleChange?.invoke()
            debouncer.cancel()
        })
    }

    private fun listRuleFiles(rootPath: String, config: AIConfig): List<File> {
        val rootDir = File(rootPath)
        val ruleRootFile = File(rootDir, config.ruleRoot)
        if (!ruleRootFile.exists()) return emptyList()

        if (config.strategy == "single-file") {
            return if (ruleRootFile.isFile) listOf(ruleRootFile) else emptyList()
        }

        val extension = config.extension?.lowercase()
        return ruleRootFile.walkTopDown()
            .filter { it.isFile }
            .filter { extension == null || it.name.lowercase().endsWith(extension) }
            .toList()
    }

    private fun collectRuleStats(rootPath: String, aiTools: Map<String, AIConfig>): List<RuleStat> {
        val allRules = mutableListOf<RuleStat>()
        for ((key, config) in aiTools) {
            val files = listRuleFiles(rootPath, config)
            files.forEach { file ->
                allRules.add(RuleStat(key, file, file.lastModified()))
            }
        }
        return allRules
    }

    private fun readRuleFiles(rootPath: String, config: AIConfig): List<RuleFile> {
        val rootDir = File(rootPath)
        return listRuleFiles(rootPath, config).map { file ->
            val relativePath = file.relativeTo(rootDir).path
            RuleFile(relativePath, file.readText(), file.lastModified())
        }
    }

    private fun getSyncState(ruleRoot: File, currentMtime: Long): SyncState? {
        val state = LocalStorage.getData(ruleRoot.absolutePath, SyncState::class.java)
        if (state != null) {
            // Check if lastModified matches
            if (state.lastModified != currentMtime) return null
        }
        return state
    }

    private fun checkAndSync() {
        if (!ConfigService.getInstance(project).getConfig().syncRules) return

        val basePath = project.basePath ?: return

        // 1. Collect all rule files and their mtimes
        val aiTools = AITool.instance.getAllAIConfigs()
        val allRules = collectRuleStats(basePath, aiTools)

        if (allRules.isEmpty()) return

        val aiMaxMtimes = allRules.groupBy { it.ai }
            .mapValues { entry -> entry.value.maxOf { it.mtime } }

        // 2. Filter candidates (ignore isFromSynced)
        val validCandidates = allRules.filter { rule ->
            val config = aiTools[rule.ai]!!
            val ruleRoot = File(basePath, config.ruleRoot)
            val mtime = aiMaxMtimes[rule.ai]!!
            val state = getSyncState(ruleRoot, mtime)
            state == null || !state.isFromSynced
        }

        if (validCandidates.isEmpty()) return

        // 3. Find the latest modified rule from valid candidates
        val latest = validCandidates.maxByOrNull { it.mtime } ?: return

        // 4. Identify current AI tool
        val currentToolKey = detectCurrentTool(project)
        if (currentToolKey == null) {
            logger.info("Current IDE '$currentAppName' is not a known AI tool target. Skipping sync.")
            return
        }

        if (latest.ai == currentToolKey) {
            return
        }

        // Check locks
        if (isLocked(basePath, latest.ai)) {
            logger.info("Sync ignored: Source ${latest.ai} is currently locked.")
            return
        }

        // Check if already synced
        val targetConfig = aiTools[currentToolKey]!!
        val targetRuleRoot = File(basePath, targetConfig.ruleRoot)
        val targetMtime = aiMaxMtimes[currentToolKey] ?: 0L

        val targetState = getSyncState(targetRuleRoot, targetMtime)

        if (targetState != null && targetState.source == latest.ai && targetMtime == latest.mtime) {
            return
        }

        // 4. Sync
        logger.info("Latest rule modification: ${latest.ai} - ${latest.file.name}")
        logger.info("Syncing rules from ${latest.ai} to $currentToolKey...")

        if (tryAcquireLock(basePath, currentToolKey)) {
            try {
                val syncedMtime = syncRules(latest.ai, currentToolKey, basePath)
                if (syncedMtime > 0) {
                    updateSyncStates(basePath, aiTools, latest, currentToolKey, aiMaxMtimes, syncedMtime)
                }
            } catch (e: Exception) {
                logger.error("Sync failed", e)
            } finally {
                releaseLock(basePath, currentToolKey)
            }
        } else {
            logger.info("Skipping sync: Could not acquire lock for $currentToolKey.")
        }
    }

    private fun detectCurrentTool(project: Project): String? {
        return AITool.instance.detectCurrentTool(project)
    }

    private fun getLockPath(rootPath: String, aiTool: String): File {
        val encodedPath = Base64.getEncoder().encodeToString(rootPath.toByteArray()).replace("/", "_").replace("+", "-")
        val homeDir = System.getProperty("user.home")
        val lockDir = File(homeDir, ".one-ide/locks")
        if (!lockDir.exists()) {
            lockDir.mkdirs()
        }
        return File(lockDir, "$encodedPath-$aiTool.lock")
    }

    private fun isLocked(rootPath: String, aiTool: String): Boolean {
        val lockFile = getLockPath(rootPath, aiTool)
        if (lockFile.exists()) {
            if (System.currentTimeMillis() - lockFile.lastModified() > 3 * 60 * 1000) {
                return false
            }
            return true
        }
        return false
    }

    private fun tryAcquireLock(rootPath: String, aiTool: String): Boolean {
        val lockFile = getLockPath(rootPath, aiTool)
        if (lockFile.exists()) {
            if (System.currentTimeMillis() - lockFile.lastModified() < 3 * 60 * 1000) {
                return false
            }
        }
        try {
            lockFile.writeText(System.currentTimeMillis().toString())
            return true
        } catch (e: Exception) {
            logger.error("Failed to acquire lock $lockFile", e)
            return false
        }
    }

    private fun releaseLock(rootPath: String, aiTool: String) {
        val lockFile = getLockPath(rootPath, aiTool)
        try {
            if (lockFile.exists()) {
                lockFile.delete()
            }
        } catch (e: Exception) {
            logger.error("Failed to release lock $lockFile", e)
        }
    }

    private fun getRuleBuilder(targetConfig: AIConfig): RuleBuilder {
        return if (targetConfig.strategy == "single-file") {
            SingleFileRuleBuilder(targetConfig.ruleRoot)
        } else {
            FolderRuleBuilder(targetConfig.ruleRoot, targetConfig.extension)
        }
    }

    private fun updateSyncStates(
        basePath: String,
        aiTools: Map<String, AIConfig>,
        latest: RuleStat,
        targetAi: String,
        aiMaxMtimes: Map<String, Long>,
        syncedMtime: Long
    ) {
        val newTargetState = SyncState()
        newTargetState.source = latest.ai
        newTargetState.lastModified = syncedMtime
        newTargetState.isFromSynced = true
        LocalStorage.setData(File(basePath, aiTools[targetAi]!!.ruleRoot).absolutePath, newTargetState)

        val sourceConfig = aiTools[latest.ai]!!
        val sourceRuleRoot = File(basePath, sourceConfig.ruleRoot)
        val sourceState = getSyncState(sourceRuleRoot, aiMaxMtimes[latest.ai] ?: 0L) ?: SyncState()
        if (!sourceState.syncedTo.contains(targetAi)) {
            sourceState.syncedTo.add(targetAi)
            LocalStorage.setData(sourceRuleRoot.absolutePath, sourceState)
        }
    }

    private fun syncRules(sourceAi: String, targetAi: String, rootPath: String): Long {
        try {
            val aiTools = AITool.instance.getAllAIConfigs()
            val sourceConfig = aiTools[sourceAi] ?: return 0L
            val targetConfig = aiTools[targetAi] ?: return 0L

            // Collect source contents
            if (!isStrategyCompatible(sourceConfig, targetConfig, rootPath)) return 0L

            val sources = readRuleFiles(rootPath, sourceConfig)

            if (sources.isEmpty()) return 0L

            // Build rules
            val builder = getRuleBuilder(targetConfig)

            val ruleFiles = builder.buildRules(sources, sourceAi)

            // Prepare revert map
            val previousContents = mutableMapOf<String, String?>()
            var maxWrittenMtime = 0L

            // We need to run write action on EDT
            var anyChanged = false
            ApplicationManager.getApplication().invokeAndWait {
                WriteAction.run<Throwable> {
                    var changed = false
                    for (ruleFile in ruleFiles) {
                        val targetFile = File(rootPath, ruleFile.path)

                        val currentContent = if (targetFile.exists()) targetFile.readText() else null
                        val newContent = ruleFile.content
                        val newMtime = ruleFile.lastModified

                        if (currentContent != newContent || targetFile.lastModified() != newMtime) {
                            previousContents[ruleFile.path] = currentContent

                            try {
                                // Using VFS to write
                                val parentDir = targetFile.parentFile
                                if (!parentDir.exists()) parentDir.mkdirs()

                                val vTargetDir = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(parentDir)

                                if (vTargetDir != null) {
                                    val vFile = vTargetDir.findChild(targetFile.name) ?: vTargetDir.createChildData(
                                        this,
                                        targetFile.name
                                    )
                                    vFile.setBinaryContent(newContent.toByteArray())
                                    logger.info("Updated ${targetFile.path}")
                                    changed = true
                                } else {
                                    // Fallback to IO
                                    targetFile.writeText(newContent)
                                    targetFile.setLastModified(newMtime)
                                    logger.info("Updated ${targetFile.path} (IO)")
                                    changed = true
                                }
                                if (newMtime > maxWrittenMtime) maxWrittenMtime = newMtime
                            } catch (e: Exception) {
                                logger.error("Failed to write ${targetFile.path}", e)
                            }
                        } else {
                            // Content and mtime match, keep track of max time
                            if (newMtime > maxWrittenMtime) maxWrittenMtime = newMtime
                        }
                    }

                    if (changed) {
                        anyChanged = true
                        showRevertNotification(rootPath, previousContents, targetAi)
                    }
                }
            }
            return if (anyChanged) maxWrittenMtime else 0L
        } catch (e: Exception) {
            logger.error("Error during syncRules", e)
            return 0L
        }
    }

    private fun isStrategyCompatible(sourceConfig: AIConfig, targetConfig: AIConfig, rootPath: String): Boolean {
        if (sourceConfig.strategy == "single-file" && targetConfig.strategy == "folder") {
            val existing = listRuleFiles(rootPath, targetConfig)
            if (existing.size > 1) {
                logger.info("Skipping sync: ${targetConfig.name} already has multiple rule files.")
                return false
            }
        }
        return true
    }

    private fun showRevertNotification(rootPath: String, previousContents: Map<String, String?>, targetAi: String) {
        val notification = NotificationGroupManager.getInstance()
            .getNotificationGroup("OneIDE Notification Group")
            .createNotification("Rules synced to $targetAi", NotificationType.INFORMATION)

        notification.addAction(NotificationAction.createSimple("Revert") {
            WriteAction.run<Throwable> {
                for ((path, content) in previousContents) {
                    val file = File(rootPath, path)
                    if (content == null) {
                        if (file.exists()) file.delete()
                    } else {
                        file.writeText(content)
                    }
                }
                notification.expire()
                logger.info("Reverted changes for $targetAi")
            }
        })

        notification.notify(project)
    }
}

File: vscode-extension/src/services/RuleService.ts
import * as vscode from 'vscode';
import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { Logger } from '../logger';
import { Debouncer } from '../utils/Debouncer';
import { IdeMetaData } from '../IdeMetaData';
import { AITool, AIConfig } from './AITool';
import { ConfigService } from './ConfigService';
import { ClusterService, RoleType } from './ClusterService';
import { LocalStorage } from './LocalStorage';

interface RuleFile {
    path: string; // relative path from rule root
    content: string;
}

interface RuleStat {
    ai: string;
    file: string;
    mtime: number;
}

interface SyncState {
    source?: string;
    lastModified: number;
    isFromSynced: boolean;
    syncedTo: string[];
}

/**
 * Interface for building rule files for a target AI tool.
 */
interface RuleBuilder {
    buildRules(sourceFiles: Array<{ path: string, content: string }>, sourceAi: string): RuleFile[];
}

/**
 * Builds rules for tools that support folder-based rule structure.
 */
export class FolderRuleBuilder implements RuleBuilder {
    constructor(private ruleRoot: string, private extension?: string) { }

    buildRules(sourceFiles: Array<{ path: string, content: string }>, sourceAi: string): RuleFile[] {
        return sourceFiles.map(file => {
            const fileName = path.basename(file.path);
            let name = fileName;

            if (this.extension) {
                const ext = path.extname(name);
                if (ext && ext !== this.extension) {
                    name = name.substring(0, name.length - ext.length) + this.extension;
                } else if (!ext) {
                    name += this.extension;
                }
            }

            const rulePath = this.ruleRoot ? path.join(this.ruleRoot, name) : name;

            return {
                path: rulePath,
                content: file.content
            };
        });
    }
}

/**
 * Builds rules for tools that require a single file configuration.
 * Concatenates all source rules into one file.
 */
export class SingleFileRuleBuilder implements RuleBuilder {
    constructor(private targetPath: string) { }

    buildRules(sourceFiles: Array<{ path: string, content: string }>, sourceAi: string): RuleFile[] {
        // Iterate through each source file and extract its content
        const mergedContent = sourceFiles
            .map(file => file.content)
            .join('\n\n');

        // Return the single target file with the merged content
        return [{
            path: this.targetPath,
            content: mergedContent ? mergedContent + '\n\n' : ''
        }];
    }
}

/**
 * Service responsible for synchronizing AI rules between different tools.
 * It monitors file changes and triggers synchronization when rules are modified.
 */
export class RuleService {
    private debouncer = new Debouncer(1000); // 1 second debounce
    private currentAppName: string;
    private configService: ConfigService;
    private clusterService: ClusterService;
    private unsubscribeRoleChange: (() => void) | undefined;

    constructor(configService: ConfigService, clusterService: ClusterService) {
        this.configService = configService;
        this.clusterService = clusterService;
        this.currentAppName = IdeMetaData.getInstance().appName;
        Logger.log(`RuleService initialized. Detected App Name: ${this.currentAppName}`);
    }

    /**
     * Starts the RuleService.
     * Triggers synchronization only when this instance becomes the cluster Leader.
     */
    public start() {
        this.unsubscribeRoleChange = this.clusterService.addRoleChangeListener((role) => {
            if (role !== RoleType.LEADER) return;
            this.debouncer.debounce(() => this.checkAndSync());
        });
    }

    public dispose() {
        this.unsubscribeRoleChange?.();
        this.debouncer.cancel();
    }

    private listRuleFiles(rootPath: string, config: AIConfig): string[] {
        const ruleRootPath = path.join(rootPath, config.ruleRoot);
        if (!fs.existsSync(ruleRootPath)) return [];

        if (config.strategy === 'single-file') {
            const stat = fs.statSync(ruleRootPath);
            return stat.isFile() ? [ruleRootPath] : [];
        }

        const results: string[] = [];
        const extension = config.extension?.toLowerCase();
        const stack = [ruleRootPath];
        while (stack.length > 0) {
            const current = stack.pop() as string;
            const entries = fs.readdirSync(current, { withFileTypes: true });
            for (const entry of entries) {
                const entryPath = path.join(current, entry.name);
                if (entry.isDirectory()) {
                    stack.push(entryPath);
                } else if (entry.isFile()) {
                    if (!extension || entry.name.toLowerCase().endsWith(extension)) {
                        results.push(entryPath);
                    }
                }
            }
        }
        return results;
    }

    private readRuleStats(aiKey: string, filePaths: string[]): RuleStat[] {
        const rules: RuleStat[] = [];
        for (const filePath of filePaths) {
            try {
                const stat = fs.statSync(filePath);
                if (stat.isFile()) {
                    rules.push({ ai: aiKey, file: filePath, mtime: stat.mtimeMs });
                }
            } catch (e) {
                Logger.error(`Error processing file ${filePath}`, e);
            }
        }
        return rules;
    }

    private readRuleFiles(filePaths: string[]): RuleFile[] {
        const sources: RuleFile[] = [];
        for (const filePath of filePaths) {
            try {
                const stat = fs.statSync(filePath);
                if (stat.isFile()) {
                    sources.push({ path: filePath, content: fs.readFileSync(filePath, 'utf-8') });
                }
            } catch (e) {
                Logger.error(`Error processing file ${filePath}`, e);
            }
        }
        return sources;
    }

    /**
     * Retrieves the synchronization state for a specific rule root.
     * 
     * @param ruleRoot The root directory of the rules.
     * @param currentMtime The current modification time to validate against.
     * @returns The SyncState if found and valid, null otherwise.
     */
    private getSyncState(ruleRoot: string, currentMtime: number): SyncState | null {
        const state = LocalStorage.getInstance().getData<SyncState>(ruleRoot);
        if (state) {
            // Check if lastModified matches
            if (state.lastModified !== currentMtime) return null;
        }
        return state;
    }

    /**
     * Checks for rule updates and synchronizes them if necessary.
     * This is the main entry point for the synchronization logic.
     */
    private async checkAndSync() {
        if (!this.configService.getConfig().syncRules) {
            Logger.log('Sync disabled in config.');
            return;
        }

        const workspaceFolders = vscode.workspace.workspaceFolders;
        if (!workspaceFolders) return;

        const rootPath = workspaceFolders[0].uri.fsPath;

        // 1. Collect all rule files and their mtimes
        const allRules: RuleStat[] = [];
        const aiTools = AITool.getInstance().getAllAIConfigs();

        for (const [aiKey, config] of Object.entries(aiTools)) {
            const filePaths = this.listRuleFiles(rootPath, config);
            allRules.push(...this.readRuleStats(aiKey, filePaths));
        }

        Logger.log(`Found ${allRules.length} rules.`);
        if (allRules.length === 0) return;

        const aiMaxMtimes = allRules.reduce<Record<string, number>>((acc, rule) => {
            acc[rule.ai] = Math.max(acc[rule.ai] ?? 0, rule.mtime);
            return acc;
        }, {});

        // 2. Filter candidates (ignore isFromSynced)
        const validCandidates = allRules.filter(rule => {
            const config = aiTools[rule.ai];
            const ruleRoot = path.join(rootPath, config.ruleRoot);
            const mtime = aiMaxMtimes[rule.ai];
            const state = this.getSyncState(ruleRoot, mtime);
            return !state || !state.isFromSynced;
        });

        Logger.log(`Valid candidates: ${validCandidates.length}`);
        if (validCandidates.length === 0) return;

        // 3. Find the latest modified rule from valid candidates
        const latest = validCandidates.reduce((prev, current) => (prev.mtime > current.mtime) ? prev : current);

        // 4. Identify current AI tool
        let currentToolKey = this.detectCurrentTool();
        Logger.log(`Detected current tool: ${currentToolKey}`);
        if (!currentToolKey) {
            Logger.log(`Current IDE '${this.currentAppName}' is not a known AI tool target. Skipping sync.`);
            return;
        }

        if (latest.ai === currentToolKey) {
            Logger.log(`Latest rule belongs to current tool (${currentToolKey}). No sync needed.`);
            return;
        }

        // Check locks
        if (this.isLocked(rootPath, latest.ai)) {
            Logger.log(`Sync ignored: Source ${latest.ai} is currently locked (active write).`);
            return;
        }

        const targetRuleRoot = path.join(rootPath, aiTools[currentToolKey].ruleRoot);
        const targetState = this.getSyncState(targetRuleRoot, aiMaxMtimes[currentToolKey] || 0);

        // Check if already synced
        if (targetState && targetState.source === latest.ai && targetState.lastModified >= latest.mtime) {
            Logger.log(`Already synced from ${latest.ai} to ${currentToolKey}.`);
            return;
        }

        // Check if source already believes it synced to us
        // This prevents re-syncing if the user manually reverts/modifies the target
        const sourceRuleRoot = path.join(rootPath, aiTools[latest.ai].ruleRoot);
        const sourceState = this.getSyncState(sourceRuleRoot, latest.mtime);

        if (sourceState && sourceState.syncedTo.includes(currentToolKey)) {
            Logger.log(`Source ${latest.ai} already synced to ${currentToolKey} (v${latest.mtime}). Skipping.`);
            return;
        }

        // 5. Sync
        Logger.log(`Latest rule modification: ${latest.ai} - ${latest.file}`);
        Logger.log(`Syncing rules from ${latest.ai} to ${currentToolKey}...`);

        if (this.tryAcquireLock(rootPath, currentToolKey)) {
            try {
                if (await this.syncRules(latest.ai, currentToolKey, rootPath)) {
                    this.updateSyncStates(rootPath, aiTools, latest, currentToolKey, targetState, aiMaxMtimes);
                }
            } finally {
                this.releaseLock(rootPath, currentToolKey);
            }
        } else {
            Logger.log(`Skipping sync: Could not acquire lock for ${currentToolKey} (recently synced).`);
        }
    }

    /**
     * Detects the currently active AI tool based on configuration or application name.
     * 
     * @returns The key of the detected AI tool, or undefined if not recognized.
     */
    private detectCurrentTool(): string | undefined {
        return AITool.getInstance().detectCurrentTool();
    }

    /**
     * Generates a unique lock file path for a given root path and AI tool.
     * 
     * @param rootPath The project root path.
     * @param aiTool The AI tool identifier.
     * @returns The absolute path to the lock file.
     */
    private getLockPath(rootPath: string, aiTool: string): string {
        const encodedPath = Buffer.from(rootPath).toString('base64').replace(/\//g, '_').replace(/\+/g, '-');
        const homeDir = os.homedir();
        const lockDir = path.join(homeDir, '.one-ide', 'locks');
        if (!fs.existsSync(lockDir)) {
            fs.mkdirSync(lockDir, { recursive: true });
        }
        return path.join(lockDir, `${encodedPath}-${aiTool}.lock`);
    }

    /**
     * Checks if a lock exists and is valid (not stale) for the given AI tool.
     * A lock is considered stale if it's older than 3 minutes.
     * 
     * @param rootPath The project root path.
     * @param aiTool The AI tool identifier.
     * @returns True if locked, false otherwise.
     */
    private isLocked(rootPath: string, aiTool: string): boolean {
        const lockFile = this.getLockPath(rootPath, aiTool);
        if (fs.existsSync(lockFile)) {
            // Check staleness (3 mins)
            const stat = fs.statSync(lockFile);
            if (Date.now() - stat.mtimeMs > 3 * 60 * 1000) {
                return false; // Stale lock
            }
            return true;
        }
        return false;
    }

    /**
     * Attempts to acquire a lock for the given AI tool.
     * 
     * @param rootPath The project root path.
     * @param aiTool The AI tool identifier.
     * @returns True if the lock was successfully acquired, false otherwise.
     */
    private tryAcquireLock(rootPath: string, aiTool: string): boolean {
        const lockFile = this.getLockPath(rootPath, aiTool);
        if (fs.existsSync(lockFile)) {
            const stat = fs.statSync(lockFile);
            if (Date.now() - stat.mtimeMs < 3 * 60 * 1000) {
                return false; // Locked
            }
            // Stale, overwrite
        }
        try {
            fs.writeFileSync(lockFile, Date.now().toString());
            return true;
        } catch (e) {
            Logger.error(`Failed to acquire lock ${lockFile}`, e);
            return false;
        }
    }

    /**
     * Releases the lock for the given AI tool.
     * 
     * @param rootPath The project root path.
     * @param aiTool The AI tool identifier.
     */
    private releaseLock(rootPath: string, aiTool: string) {
        const lockFile = this.getLockPath(rootPath, aiTool);
        try {
            if (fs.existsSync(lockFile)) {
                fs.unlinkSync(lockFile);
            }
        } catch (e) {
            Logger.error(`Failed to release lock ${lockFile}`, e);
        }
    }

    private getRuleBuilder(targetConfig: AIConfig): RuleBuilder {
        if (targetConfig.strategy === 'single-file') {
            return new SingleFileRuleBuilder(targetConfig.ruleRoot);
        } else {
            return new FolderRuleBuilder(targetConfig.ruleRoot, targetConfig.extension);
        }
    }

    private updateSyncStates(
        rootPath: string,
        aiTools: Record<string, AIConfig>,
        latest: RuleStat,
        targetAi: string,
        targetState: SyncState | null,
        aiMaxMtimes: Record<string, number>
    ) {
        const targetRuleRoot = path.join(rootPath, aiTools[targetAi].ruleRoot);
        const newTargetState: SyncState = {
            lastModified: latest.mtime,
            isFromSynced: true,
            source: latest.ai,
            syncedTo: targetState?.syncedTo || []
        };
        LocalStorage.getInstance().setData(targetRuleRoot, newTargetState);

        const sourceRuleRoot = path.join(rootPath, aiTools[latest.ai].ruleRoot);
        const sourceState = this.getSyncState(sourceRuleRoot, aiMaxMtimes[latest.ai]) || {
            lastModified: latest.mtime,
            isFromSynced: false,
            source: undefined,
            syncedTo: []
        };

        if (!sourceState.syncedTo.includes(targetAi)) {
            sourceState.syncedTo.push(targetAi);
        }
        LocalStorage.getInstance().setData(sourceRuleRoot, sourceState);
    }

    /**
     * Synchronizes rules from the source AI tool to the target AI tool.
     * 
     * @param sourceAi The identifier of the source AI tool.
     * @param targetAi The identifier of the target AI tool.
     * @param rootPath The project root path.
     * @returns A promise that resolves to true if changes were made, false otherwise.
     */
    private async syncRules(sourceAi: string, targetAi: string, rootPath: string): Promise<boolean> {
        const aiTools = AITool.getInstance().getAllAIConfigs();
        const sourceConfig = aiTools[sourceAi];
        const targetConfig = aiTools[targetAi];

        if (!sourceConfig || !targetConfig) return false;

        if (!this.isStrategyCompatible(sourceConfig, targetConfig, rootPath)) {
            return false;
        }

        const filePaths = this.listRuleFiles(rootPath, sourceConfig);
        const sources = this.readRuleFiles(filePaths);

        if (sources.length === 0) return false;

        // Build rules based on strategy
        const builder = this.getRuleBuilder(targetConfig);

        const ruleFiles = builder.buildRules(sources, sourceAi);

        // Write files
        const previousContents: Record<string, string | null> = {};
        let changed = false;

        for (const ruleFile of ruleFiles) {
            const targetPath = path.join(rootPath, ruleFile.path);

            try {
                let currentContent: string | null = null;
                if (fs.existsSync(targetPath)) {
                    currentContent = fs.readFileSync(targetPath, 'utf-8');
                }

                if (currentContent !== ruleFile.content) {
                    previousContents[ruleFile.path] = currentContent;

                    const targetDir = path.dirname(targetPath);
                    if (!fs.existsSync(targetDir)) {
                        fs.mkdirSync(targetDir, { recursive: true });
                    }

                    fs.writeFileSync(targetPath, ruleFile.content);
                    Logger.log(`Updated ${targetPath}`);
                    changed = true;
                }
            } catch (e) {
                Logger.error(`Failed to write ${targetPath}`, e);
            }
        }

        if (changed) {
            this.showRevertNotification(rootPath, previousContents, targetAi);
        }
        return changed;
    }

    private isStrategyCompatible(sourceConfig: AIConfig, targetConfig: AIConfig, rootPath: string): boolean {
        if (sourceConfig.strategy === 'single-file' && targetConfig.strategy === 'folder') {
            const existing = this.listRuleFiles(rootPath, targetConfig);
            if (existing.length > 1) {
                Logger.log(`Skipping sync: ${targetConfig.name} already has multiple rule files.`);
                return false;
            }
        }
        return true;
    }

    private showRevertNotification(rootPath: string, previousContents: Record<string, string | null>, targetAi: string) {
        vscode.window.showInformationMessage(`Rules synced to ${targetAi}`, 'Revert')
            .then(selection => {
                if (selection === 'Revert') {
                    try {
                        for (const [relativePath, content] of Object.entries(previousContents)) {
                            const targetPath = path.join(rootPath, relativePath);
                            if (content === null) {
                                if (fs.existsSync(targetPath)) {
                                    fs.unlinkSync(targetPath);
                                }
                            } else {
                                const targetDir = path.dirname(targetPath);
                                if (!fs.existsSync(targetDir)) {
                                    fs.mkdirSync(targetDir, { recursive: true });
                                }
                                fs.writeFileSync(targetPath, content);
                            }
                        }
                        Logger.log(`Reverted changes for ${targetAi}`);
                    } catch (e) {
                        Logger.error(`Failed to revert changes for ${targetAi}`, e);
                        vscode.window.showErrorMessage(`Failed to revert changes: ${e}`);
                    }
                }
            });
    }
}
